### ISSUE TITLE
[3.1] Postgresql bug. Invalid optimization.

### ISSUE THREAD

**Issue by grehovm** [schotime/NPoco#259]
NPoco use invalid optimization. Parameter `password` has type `character varying`. Parameter `script` has type `text`. Should not be used one parameter for a different types, even when the value is null.
`NpgsqlException: inconsistent types deduced for parameter character varying and text`
```
UPDATE "converter" SET "executorid" = @0, "name" = @1, "filename" = @2, "protection" = @3, "password" = @4, "script" = @4, "created" = @5, "updated" = @6, "archive" = @3 WHERE "id" = @7
	 -> @0 [Guid] = "f1ff5322-35ed-4c7d-90bd-dea65df12a73"
	 -> @1 [String] = "asdas1231"
	 -> @2 [String] = "asdasd1212"
	 -> @3 [Boolean] = "False"
	 -> @4 [DBNull] = ""
	 -> @5 [DateTime] = "11.04.2016 3:41:11"
	 -> @6 [DateTime] = "11.04.2016 4:30:30"
	 -> @7 [Guid] = "ebf406cd-807f-40ea-81be-1ae9a12bc850"
```
Workaround in mappings?

**Comment by grehovm** [comment]
I did an investigation and found where there is a replacement. [Here](https://github.com/schotime/NPoco/blob/master/src/NPoco/ParameterHelper.cs#L96-L101)
Previously encountered problems in the ParameterHelper due to the lack of support for arrays postgresql. Npgsql supports passing arrays in one parameter.

**Comment by grehovm** [comment]
Incidentally db.Insert(user); does not use this optimization.
```
INSERT INTO "user" ("id","companyid","name","passwordhash","securitystamp","email","emailconfirmed","normalizedemail","accessfailedcount","lockoutenabled","lockoutend","enabled") VALUES (@0,@1,@2,@3,@4,@5,@6,@7,@8,@9,@10,@11)
	 -> @0 [Guid] = "569b365a-c754-4635-87d5-0eff6165f6e0"
	 -> @1 [Guid] = "feee1703-0aed-41aa-8a2f-00299dc31e0b"
	 -> @2 [String] = "##TEST##"
	 -> @3 [DBNull] = ""
	 -> @4 [String] = "asdasd"
	 -> @5 [String] = "email@email"
	 -> @6 [Boolean] = "False"
	 -> @7 [String] = "asdasd"
	 -> @8 [Int32] = "0"
	 -> @9 [Boolean] = "True"
	 -> @10 [DBNull] = ""
	 -> @11 [Boolean] = "True"
```
db.Update(user);
```
UPDATE "user" SET "companyid" = @0, "name" = @1, "passwordhash" = @2, "securitystamp" = @3, "email" = @4, "emailconfirmed" = @5, "normalizedemail" = @3, "accessfailedcount" = @6, "lockoutenabled" = @7, "lockoutend" = @2, "enabled" = @7 WHERE "id" = @8
	 -> @0 [Guid] = "feee1703-0aed-41aa-8a2f-00299dc31e0b"
	 -> @1 [String] = "##TEST##"
	 -> @2 [DBNull] = ""
	 -> @3 [String] = "asdasd"
	 -> @4 [String] = "email@email"
	 -> @5 [Boolean] = "False"
	 -> @6 [Int32] = "0"
	 -> @7 [Boolean] = "True"
	 -> @8 [Guid] = "569b365a-c754-4635-87d5-0eff6165f6e0"
```
For strings enough to use `string.empy` as a workaround. But for `Nullable<T>` it's completely broken. Two `Nullable<T>` or `Nullable<T>` and string(null) will give us an error. As I understand it, this error does not allow using NPoco with Postgresql. Please, help. I blocked by this issue.

**Comment by lerrand** [comment]
I am experiencing something very similar with NPoco 3.1.0 and Npgsql 3.0.5 (parameter inconsistency is `character`  vs ` timestamp with timezone` though).

Shot in the dark here but maybe it could have something to do with some weird mishandling of `DBNull`'s due to the fact that Npgsql has changed the way it handles [null values and other types](http://www.npgsql.org/doc/migration-3.0.html) with their v3.x releases? (see "Type Handling" section of linked page). NPoco 3.1.0 works fine with Npgsql 2.2.7 by the way.

**Comment by grehovm** [comment]
Perhaps earlier Npgsql smooth out Postgresql's conditions. But this does not negate the fact that NPoco generates invalid sql for Postgresql. An error occurs because Postgresql does not allow cast one parameter to different types.

**Comment by lerrand** [comment]
Oh I have no doubt this is an NPoco issue! What I was trying with my previous comment was to give some extra bugshooting info to @schotime 

Just to clarify: my theory is that NPoco is sending `null`-valued parameters to Npgsql when, according to the Npgsql v3 migration docs, `DBNull.value` should be sent. This would probably explain the bad SQL (as a direct consequence of NPoco sending bad input values to Npgsql)?

**Comment by schotime** [comment]
We convert null to dbnull.value in AddParameter(). That won't be it.

**Comment by grehovm** [comment]
I watched source. To insert npoco creates sql and then executes through internal method. To update npoco creates sql then executes it through the public method 'int Execute (string sql, params object [] args)' which makes the optimization of the parameters and for parameters with the same value leaves only one object. The error occurs in Postgsql when he can not single parameter DBNull cast to different types (text, timestamp, character varying).

**Comment by schotime** [comment]
Can this optimization be made for queries? Can you verify?

**Comment by grehovm** [comment]
I do not understand what you want me to check.
Execute function [get sql](https://github.com/schotime/NPoco/blob/master/src/NPoco/Database.cs#L737) wich [builds in getter](https://github.com/schotime/NPoco/blob/master/src/NPoco/Sql.cs#L60) and during the build call [ParameterHelper.ProcessParams(...)](https://github.com/schotime/NPoco/blob/master/src/NPoco/Sql.cs#L117) which in turn [makes optimization](https://github.com/schotime/NPoco/blob/master/src/NPoco/ParameterHelper.cs#L96-L101)

**Comment by schotime** [comment]
Can you run a select statement with the two values being the same?

**Comment by grehovm** [comment]
Select
```
var users = db.Fetch<User>(@"select * from ""user"" where lockoutend=@0 and passwordhash=@1", null, null);
```
LastCommand:
```
select * from "user" where lockoutend=@0 and passwordhash=@0
	 -> @0 [DBNull] = ""
```
Exception:
42883: operator does not exist: character varying = timestamp with time zone

**Comment by schotime** [comment]
Cool.
I wonder if it's just nulls?

**Comment by grehovm** [comment]
No. I received an error for strings.
'inconsistent types deduced for parameter character varying and text'
Any objects that have the same value in the C# and various types DB

**Comment by schotime** [comment]
Fixed in 3.2
